lazy_tensor_core/lazy_tensor_core/csrc/tensor_impl.cpp

```diff
@@ -10,31 +10,40 @@
 namespace torch_lazy_tensors {
 namespace {
 
+// LTCGuardImpl is used by CompositeExplicitAutograd ops or eager fallbacks to make sure that some particular tensors
+// within the life scope of the guard are on the same device. For example, in RegisterCompositeExplicitAutograd.cpp,
+// outputs of each op are examined if they are on same device as the supplied TensorOptions. For more information,
+// see DeviceGuard.h.
+// For ops that have LTC native function implementations, this guard is omitted.
+thread_local c10::Device g_device(c10::DeviceType::Lazy);
+
 struct LTCGuardImpl : public c10::impl::DeviceGuardImplInterface {
   at::DeviceType type() const override { return at::DeviceType::Lazy; }
 
   c10::Device exchangeDevice(c10::Device device) const override {
-    return bridge::SetCurrentDevice(device);
+    auto old_device = g_device;
+    g_device = device;
+    return old_device;
   }
 
   c10::Device getDevice() const override {
-    return bridge::GetCurrentAtenDevice();
+    return g_device;
   }
 
   void setDevice(c10::Device device) const override {
-    bridge::SetCurrentDevice(device);
+    g_device = device;
```

Comment:
do we need to sanity check here? (can someone set a non-lazy device type?)

Reply:
Let me add a debug assert. It seems unlikely.